### ISSUE TITLE
Not working in build (tried prior fixes)

### ISSUE THREAD

**Issue by smoore2171** [Ayfel/PrefabLightmapping#33]
I have an older project on 2019.3 I'm trying to get to work with this script.

This works fine while in editor, but in the built game, some nested prefabs using the unity Standard shader that are already baked using realtime GI in the scene show up as unlit as if the lightmap is not applied, both in the prefab being spawned and in the scene that exists (it is the same inner prefab).

This happens whether the prefab was in the scene or spawned (it will go dark as soon as I spawn it). It happens whether or not the objects in the prefab are flagged as static or not. And it happens whether or not I strip all shaders or not (set to custom with everything checked, etc) in settings.

The only thing I notice is the textures are not under resource/*. I'm not sure if they need to be. The other level with a reference to this prefab is not included with the build. I will check just including that to see.

**Comment by smoore2171** [comment]
Should note other nested prefabs within the prefab do appear to look fine using another shader, it's just this one on the standard shader for some reason.

**Comment by Ayfel** [comment]
I am not understanding quite right the setup.  You have baked lighting or real-time GI? Are you in the built in renderer or another scriptable render pipeline? Are you saying parts of the prefab are correctly shown as baked an others not? Do try including the level you bake them in see if that changes and let me know of the results and the above question pls. Thank you.

**Comment by smoore2171** [comment]
so I have a prefab A containing sub prefab B. I place A in an empty scene with script in the root object, have it only do blaked lighting with a directional light set to baked.

I bake the lighting and use the script. So far so good.

I go another scene that contains prefab B but lit with dynamic GI (I was using to see the difference in reauthor the baked lighting to make it close to my existing realtime GI lighting) and either place prefab A in it or spawn it at runtime.

In the editor play, the lighting for everything looks correct. If I build the solution, prefab B looks like it gets its lightmap stripped everywhere, both in Prefab A and in the scene in which it is spawned. The last time I tried this, the whole prefab A actually looks like it is not correctly lit as well.

**Comment by smoore2171** [comment]
Including the level did not seem to help.

**Comment by smoore2171** [comment]
and it is built-in renderer*

**Comment by smoore2171** [comment]
I just tested a build of just this prefab A placed in that empty level, and the same behavior is happening so I don't think it is related to the dynamic GI.

**Comment by smoore2171** [comment]
I side-by-sided it with a build where the lighting wasn't built at all, and it does seem only the Standard shader is affected.. Other objects in the scene with different shaders are actually getting the lightmaps applied correctly.

**Comment by smoore2171** [comment]
I think I've narrowed it down to an existing realtime-only directional light (in addition to the baked-only realtime light) in both scenes. In the editor, these objects, even though they have baked lighting, are still receiving and calculating light from the realtime directional light. For some reason, only for builds, the Standard shader does not take realtime lighting if it has baked lightmaps. I can only imagine this inner prefab is using a shared material or something and it is applying this baked lightmap globally, which explains why the existing objects in the scene get affected.

I suppose I can just have it use the other shader if there isn't a workaround for that.

**Comment by smoore2171** [comment]
Switching to another shader that wraps Standard works. I don't know why, but at least I am unblocked

**Comment by Ayfel** [comment]
Im glad you have a workaround. It is quite a complex setup so I would love to understand the issue and see why it happens, if you could share a project that reproduced the issue I'd be happy to take a look

**Comment by smoore2171** [comment]
I'll try to repro in 2021.1 with a simple cube prefab

**Comment by Ayfel** [comment]
Thanks 2019 is fine too as it is LTS

**Comment by smoore2171** [comment]
So I narrowed this down further. It appears there was a bug in just 2019.X versions where detail albedo secondary masks for a material will get stripped in the build for standard shader if using baked lightmaps.

I tested the exact same scenario on a 2019.3 project and a 2021.X project with just a cube prefab and a new standard shader material using a built-in texture as the secondary detail albedo. It gets stripped in the builds for 2019.X but not in the 2021.X, so this issue must have been fixed somewhere along the way. This was making my object look much darker as if it was not getting lightmaps at all, but I confirmed it looked the same as the object without the detail map, which was slightly distinct from completely unlit.

So not an issue with this script it appears, just unity being unity.

**Comment by smoore2171** [comment]
I spoke too soon. It appears if I do a normal bake and load the scene with this prefab, the detail mask is kept in the built game. Only if I bake those lightmaps to this prefab does the detail map get dropped. Not really sure how the textures in the material are being modified by this script.

**Comment by smoore2171** [comment]
It's this section. There must have been some side effect to setting the shader in 2019.3 that they resolve in 2021...
` // You have to release shaders.
            Material[] mat = info.renderer.sharedMaterials;
            for (int j = 0; j < mat.Length; j++)
            {
                if (mat[j] != null && Shader.Find(mat[j].shader.name) != null)
                    mat[j].shader = Shader.Find(mat[j].shader.name);
            }`

**Comment by smoore2171** [comment]
Idk if you want to wrap that in unity version preprocessors, as I'm not clear on what resetting the material shaders was doing. It claims it is releasing shaders, but idk how the pool of shaders the "Shader.Find(" pulls differs from the shader it was pointing to at load.

**Comment by Ayfel** [comment]
So it works fine when removing the section you pointed out?

**Comment by smoore2171** [comment]
Correct, but only on 2019.3. On 2021, it worked with and without reloading the shader.

**Comment by Ayfel** [comment]
Yeah confirmed, this is a Unity issue, the problem is that in some cases I encountered the shader reload is necessary so its not just wrapping it in a preprocessor directive that will fix it, I have to think on it a bit, maybe an option in the script.

**Comment by smoore2171** [comment]
One somewhat hacky workaround is to simply store off all of the material inputs and manually send them back into the shader after reloading it.
